torch/distributed/hooks.py

```diff
@@ -0,0 +1,174 @@
+    "register_collective_end_hook",
+    "register_process_group_hook",
+]
+
+
+@dataclass
+class CollectiveStatus:
+    pg_name: str = "unknown"  # This name matches the one informed in the pgreg cb
+    backend: str = "unknown"  # Name of the backend used
+    sequence_number: int = -1  # This name matches the one informed in the pgreg cb
+    operation: str = "unknown"  # collective name
+    timestamp: int = 0  # timestamp to the earliest time we noticed this event
+    duration: Optional[float] = None  # value in milliseconds it took executing
+    drop_count: int = 0  # number of events dropped following this one
+
+
+COLLECTIVE_HOOK_TYPE = Callable[[CollectiveStatus], None]
+PG_HOOK_TYPE = Callable[[dist.ProcessGroup, str], None]
+
+# This controls the number of internal failures we'll tolerate before giving up
+_MAX_INTERNAL_FAILURES = 10
+
+"""
+TODO:
+    hook timing
+
+"""
+logger = logging.getLogger(__name__)
+_cb_thread: Optional[threading.Thread] = None
+_start_callbacks: List[COLLECTIVE_HOOK_TYPE] = []
+_end_callbacks: List[COLLECTIVE_HOOK_TYPE] = []
+_pp_r = -1
+_pp_w = -1
+
+
+def _c10d_pg_hooks_loops():
+    internal_failures = 0
+    while True:
+        # we don't care about the result, this is how we implement notification
+        _ = os.read(_pp_r, 1)
+        evt: Dict[str, object] = _dequeue_c10d_event()
+        try:
+            event_kind = evt.pop("event_kind", None)
+            if event_kind is None:
+                logger.warning(
+                    "c10d returned event dictionary %s without 'event_kind' key, cannot dispatch",
+                    evt,
+                )
+                internal_failures += 1
+                if internal_failures >= _MAX_INTERNAL_FAILURES:
+                    logger.warning(
+                        "too many internal c10d failures processing callback loop. stopping"
+                    )
+                    return
+                time.sleep(1)
+                continue
+
+            if event_kind == int(EventKind.START):  # type: ignore[call-overload]
+                cb_list = _start_callbacks
+            elif event_kind == int(EventKind.END):  # type: ignore[call-overload]
+                cb_list = _end_callbacks
+            else:
+                logger.warning(
+                    "c10d event %s with invalid 'event_kind' with value %d",
+                    evt,
+                    event_kind,
+                )
+                internal_failures += 1
+                if internal_failures >= _MAX_INTERNAL_FAILURES:
+                    logger.warning(
+                        "too many internal c10d failures processing callback loop. stopping"
+                    )
+                    return
+                time.sleep(1)
+                continue
+
+            status = CollectiveStatus(**evt)  # type: ignore[arg-type]
+            for cb in cb_list:
+                try:
+                    cb(status)
+                except Exception as e:
```

Comment:
pros/cons of leaving cb exception outside of internal_failures counter?

Reply:
Pros: when there are more than one registered callback, they won't interfere with each other (the first failing won't mess with the second).

Cons: This might be spammy.